internal/frr/frr.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package frr
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/metallb/frrk8s/internal/logging"
+)
+
+type ConfigHandler interface {
+	ApplyConfig(config *Config) error
+}
+
+type FRR struct {
+	reloadConfig chan reloadEvent
+	logLevel     string
+	sync.Mutex
+}
+
+// Create a variable for os.Hostname() in order to make it easy to mock out
+// in unit tests.
+var osHostname = os.Hostname
+
+func (f *FRR) ApplyConfig(config *Config) error {
+	hostname, err := osHostname()
+	if err != nil {
+		return err
+	}
+
+	// TODO add internal wrapper
+	config.Loglevel = f.logLevel
+	config.Hostname = hostname
+	f.reloadConfig <- reloadEvent{config: config}
+	return nil
+}
+
+var debounceTimeout = 3 * time.Second
+var failureTimeout = time.Second * 5
+
+func NewFRR(ctx context.Context, logger log.Logger, logLevel logging.Level) *FRR {
+	res := &FRR{
+		reloadConfig: make(chan reloadEvent),
+		logLevel:     logLevelToFRR(logLevel),
+	}
+	reload := func(config *Config) error {
+		return generateAndReloadConfigFile(config, logger)
+	}
+
+	debouncer(ctx, reload, res.reloadConfig, debounceTimeout, failureTimeout, logger)
+	reloadValidator(ctx, logger, res.reloadConfig)
+	return res
+}
+
+func reloadValidator(ctx context.Context, l log.Logger, reload chan<- reloadEvent) {
+	var tickerIntervals = 30 * time.Second
+	var prevReloadTimeStamp string
+
+	ticker := time.NewTicker(tickerIntervals)
+	go func() {
+		select {
+		case <-ticker.C:
+			validateReload(l, &prevReloadTimeStamp, reload)
+		case <-ctx.Done():
+			return
+		}
+	}()
+}
+
+const statusFileName = "/etc/frr_reloader/.status"
+
+func validateReload(l log.Logger, prevReloadTimeStamp *string, reload chan<- reloadEvent) {
+	bytes, err := os.ReadFile(statusFileName)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			level.Error(l).Log("op", "reload-validate", "error", err, "cause", "readFile", "fileName", statusFileName)
+		}
+		return
+	}
+
+	lastReloadStatus := strings.Fields(string(bytes))
+	if len(lastReloadStatus) != 2 {
+		level.Error(l).Log("op", "reload-validate", "error", err, "cause", "Fields", "bytes", string(bytes))
```

Comment:
we need to re-write this one. err is nil here. it's more of an invalid format error.